packages/vendure-plugin-selectable-gifts/src/api/gift.service.ts

```diff
@@ -47,7 +50,30 @@ export class GiftService {
     if (!variantIds.length) {
       return [];
     }
-    return await this.variantService.findByIds(ctx, variantIds);
+    const variants = await this.connection
+      .getRepository(ctx, ProductVariant)
+      .createQueryBuilder('variant')
+      .leftJoin('variant.stockLevels', 'stockLevel')
+      .leftJoin('variant.translations', 'translations')
+      .leftJoinAndSelect('variant.productVariantPrices', 'productVariantPrices')
+      .leftJoinAndSelect('variant.taxCategory', 'taxCategory')
+      .leftJoin('variant.channels', 'channel')
+      .addGroupBy('variant.id')
+      .addSelect(['SUM(stockLevel.stockOnHand) as stockOnHand'])
+      .addSelect(['SUM(stockLevel.stockAllocated) as stockAllocated'])
+      .where('variant.id IN (:...ids)', { ids: variantIds })
+      .andWhere('variant.enabled = true')
+      .andWhere('stockOnHand > stockAllocated')
+      .andWhere('variant.deletedAt IS NULL')
+      .andWhere('channel.id = :channelId', { channelId: ctx.channelId })
+      .getMany();
+    const translatedVariants = variants.map((v) =>
+      this.translatorService.translate(v, ctx)
+    );
+    const translatedVariantsWithPrice = translatedVariants.map((v) =>
+      this.priceApplicator.applyChannelPriceAndTax(v, ctx)
+    );
+    return Promise.all(translatedVariantsWithPrice);
```

Comment:
I think awaiting both promises this way could cause race conditions:
(Simplified example)
1. Example A is `{id: 1}`
2. Variant translation is started, but not awaited (so in progress)
3. Simultaneously, the applyPrices is being started, also not awaited
4. Translation is finished, so it will set the variant object now to `{ id: 1, name: 'Some translated name'}`
5. ApplyPrices is now also finished, but it sets the object to `{ id: 1, price: 1234}`. Notice that the translated name is now gone, because the input of applyPrices was the variant without translated fields.

We can't be 100% sure this happens without knowing the functions internals, but its best to `Promise.all` all translations first before applying prices.

Reply:
I don't think `translatorService.translate` returns a promise. It uses [translateDeep](https://github.com/vendure-ecommerce/vendure/blob/8a3b5f818046c0ae83099781d3763f17783f376e/packages/core/src/service/helpers/utils/translate-entity.ts#L96) and I think it returns a `Translated<T>`

Reply:
I misread the last line as `return Promise.all(translatedVariantsWithPrice, translatedVariants);`!

You are totally right, my bad. Well so far for my story about race conditions :laughing: